File: Review-04-Unit-Testing/Example-2-Java/src/main/java/edu/odu/cs/cs330/examples/Player.java
package edu.odu.cs.cs330.examples;

import java.io.IOException;

/**
 * This is more a Player interface than a Player class.
 * <p>
 * However, such distinctions and discussions belong in
 * the OOP and Inheritance Modules
 */
class Player implements Cloneable
{

    /**
     * Name of a player.
     */
    private String name;

    /**
     * Symbol to be used when making moves. This may change between games.
     */
    private char   symbol;

    /**
     * A Player that serves as a sentinal value or placeholder.
     */
    public static final Player REFERENCE_CYLON = new Player();

    /**
     * Checks whether a player is a placeholder or
     * an actual player.
     *
     * @param possibleCylon the player whose humanity is in question
     *
     * @return true if the player is a Cylon
     */
    public static boolean isGeneric(final Player possibleCylon)
    {
        return possibleCylon.equals(Player.REFERENCE_CYLON);
    }

    /**
     * Create a Player with a Generic name.
     */
    public Player()
    {
        name = "I. C. Generic";
        symbol = '?';
    }

    /**
     * Create a Player with a selected name.
     *
     * @param n desired name
     */
    public Player(String n)
    {
        name = n;
        symbol = '?'; // testing caught this
    }

    /**
     * Retrieve name.
     *
     * @return player name
     */
    public String getName()
    {
        return name;
    }

    /**
     * Set player name.
     *
     * @param n new name
     *
     * @pre (n.size() > 0)
     */
    public void setName(String n)
    {
        name  = n;
    }

    /**
     * Retrieve the next move.
     *
     * @return board cell id representing the selected move
     *
     * @throws IOException if the move can not be retreived from the player.
     */
    public int nextMove(Strategy theStrategy)
        throws IOException
    {
        return theStrategy.nextMove();
    }

    /**
     * Is this a Human Player?
     * <p>
     * In this discussion, always yes :(
     *
     * @return true if the player is a human
     */
    public boolean isHuman()
    {
        return true;
    }

    /**
     * Is this a Computer Player?
     * <p>
     * In this discussion, always no :(
     *
     * @return true if the player is a Cylon
     */
    public boolean isComputer()
    {
        return false;
    }

    /**
     * Retrieve player symbol to be used
     * for marking moves.
     *
     * @return current player symbol
     */
    public char getSymbol()
    {
        return symbol;
    }

    /**
     * Change the player symbol.
     *
     * @param newSymbol new character to be used by the player
     */
    public void setSymbol(char newSymbol)
    {
        symbol = newSymbol;
    }

    @Override
    public boolean equals(Object rhs)
    {
        if (!(rhs instanceof Player)) {
            return false;
        }

        Player rhsP = (Player) rhs;

        return this.name.equals(rhsP.name);
    }

    @Override
    public int hashCode()
    {
        return this.name.hashCode();
    }

    /**
     * Generate a player string, but only the name.
     */
    @Override
    public String toString()
    {
        return this.getName();
    }

    @Override
    public Player clone()
    {
        Player cpy = new Player(this.name);

        cpy.symbol = this.symbol;

        return cpy;
    }
}

File: Review-15-Python-Strategy-Pattern/Example-1/examples/player.py
from examples.strategy import Strategy


class Player:
    """
    This is more a Player interface than a Player class.
    <p>
    However, such distinctions and discussions belong in
    the OOP and Inheritance Modules
    """

    @staticmethod
    def is_generic(possible_cylon: "Player") -> bool:
        """
        Checks whether a player is a placeholder or
        an actual player.

        Args:
            possible_cylon (Player): player whose humanity is in question

        Returns:
            True if the player is a Cylon
        """

        # print(REFERENCE_CYLON)
        return possible_cylon == REFERENCE_CYLON

    def __init__(self, n: str = "I. C. Generic"):
        """
        Create a Player with a selected name.

        Args:
            n: desired name
        """

        self._name = n
        self._symbol = "?"  # testing caught this

    def get_name(self) -> str:
        """
        Retrieve name.

        Returns:
            player name
        """

        return self._name

    def set_name(self, n: str):
        """
        Set player name.

        @param n new name

        @pre (n.size() > 0)
        """

        self._name = n

    def next_move(self, the_strategy: Strategy) -> int:
        """
        Retrieve the next move.

        @return board cell id representing the selected move

        @throws IOException if the move can not be retreived from the player.
        """

        return the_strategy.next_move()

    def is_human(self) -> bool:
        """
        Is this a Human Player?

        In this discussion, always yes :(

        Returns:
            True if the player is a human
        """

        return True

    def is_computer(self):
        """
        Is this a Computer Player?

        In this discussion, always no :(

        Returns:
            True if the player is a Cylon
        """

        return False

    def get_symbol(self) -> str:
        """
        Retrieve player symbol to be used
        for marking moves.

        Returns:
            current player symbol
        """

        return self._symbol

    def set_symbol(self, new_symbol: str):
        """
        Change the player symbol.

        Args:
            new_symbol: new character to be used by the player
        """

        self._symbol = new_symbol

    def __eq__(self, rhs):
        if not isinstance(rhs, self.__class__):
            return False

        return self._name == rhs._name

    def __hash__(self):
        return hash(self._name)

    def __str__(self):
        """
        Generate a player string, but only the name.
        """

        return self._name

    def __deepcopy__(self, memo):
        """
        Create a new duplicate Player.
        """

        cpy = Player(self._name)
        cpy.set_symbol(self._symbol)

        return cpy


REFERENCE_CYLON = Player()
"""
A Player that serves as a sentinal value or placeholder.
"""
